doc/examples/nn_regression/nn_regression.cpp

```diff
@@ -0,0 +1,409 @@
+#include <mlpack/methods/ann/ffn.hpp>
+#include <mlpack/methods/ann/loss_functions/mean_squared_error.hpp>
+#include <mlpack/prereqs.hpp>
+
+using namespace arma;
+using namespace mlpack;
+using namespace mlpack::ann;
+using namespace mlpack::optimization;
+
+//TODO(zso): this should go into mlpack/core/data/feature_normalization.hpp
+namespace mlpack {
+	namespace data {
+
+		/** 
+		* \class FeatureNormalization
+		* \brief Data normalization extension module for mlpack.
+		* 
+		* \details Normalization of all features in the input data into the same range provides in most of the cases
+	    * a much better accuracy and smaller and faster models! Subsequent normalizations of new data for the same model
+		* must use the same min, max vectors and therefore they are saved in the FeatureNormalization class! You can also
+		* serialize the FeatureNormalization object and reconstruct the min, max vectors e.g. for later inference/prediction.
+		*
+		* @author Zoltan Somogyi
+		*/
+		class FeatureNormalization
+		{
+		public:
+			/** Constructor */
+			FeatureNormalization(double scalelower, double scaleupper, arma::vec & _itemMin, arma::vec & _itemMax)
+				: scale_upper(scaleupper), scale_lower(scalelower), itemMin(&_itemMin), itemMax(&_itemMax) {}
+
+			/**
+			* \fn Normalize
+			*
+			* Normalize the data into the range provided by the scale_lower and scale_upper parameters 
+			* by using the itemMin and itemMax vectors or if they do not exist yet then calculate 
+			* itemMin and itemMax first for the input data.
+			* The operation is in place (modifies the input data).
+			*
+			* @param data Input dataset
+			* @see arma::mat in mlpack documentation
+			*/
+			void Normalize(arma::mat& data)
+			{
+				// Use the itemMin, itemMax vectors if provided (e.g. for subsequent normalization with the same range)
+				// or make them if not provided
+				if (itemMin->size() < 1 || itemMax->size() < 1) {
+					*itemMin = arma::vec(data.n_rows, arma::fill::zeros);
+					*itemMax = arma::vec(data.n_rows, arma::fill::zeros);
+					for (int i = 0; i < data.n_rows; i++)
+					{
+						(*itemMin)(i) = arma::min(data.row(i));
+						(*itemMax)(i) = arma::max(data.row(i));
+					}
+				}
+
+				data.each_col([&](arma::vec& datapoint)
+				{
+					for (int i = 0; i < datapoint.size(); i++)
+					{
+						// Scale the data into the appropriate range given by scale_upper and scale_lower. 
+						// Each column (feature) in the data file has a minimum and maximum value which 
+						// are used to scale the data.
+						double Dx = scale_upper - scale_lower;
+						double val = datapoint(i);
+						// skip single-valued attribute
+						if ((*itemMin)(i) != (*itemMax)(i)) {
+							if (val == (*itemMin)(i))
+								val = scale_lower;
+							else if (val == (*itemMax)(i))
+								val = scale_upper;
+							else
+								val = scale_lower + Dx * (val - (*itemMin)(i)) / ((*itemMax)(i) - (*itemMin)(i));
+						}
+						datapoint(i) = val;
```

Comment:
What about an implementation like this?

```
datapoint = arma::clamp((datapoint - itemMin) ./ ranges, 0.0, 1.0) * Dx + scale_lower;
```

Basically this vectorized expression requires us to have a precomputed vector `ranges = itemMax - itemMin`.  This should be faster to run too, since Armadillo should be able to make use of SIMD instructions and loop unrolling where possible.

Reply:
Ryan, I am not yet very much used to armadillo, would you mind to change the code as you think that it is faster when you have some spare time (and please test it also, my code is tested and it works well)? Of course add your name as co-author.
My only remark is that do not use the clamp function because that would cause an error in **subsequent** normalization (the values may be outside the 0-1 limit.)!

Reply:
Sure, I can go over and rewrite the Armadillo expressions.  I see what you mean about the clamp; I actually misread this as `if (val <= (*itemMin)(i))`, but I see the code is different now.  (Actually I think in this case it would be equivalently fast or faster than the current implementation to simply remove the `if`/`else if`/`else` and just use the body of the `else`.  But in nay case a one-liner with Armadillo should be nicer for maintenance so we can go with that.)

Reply:
OK, just be careful with the division by 0 (I do not see yet how you can prevent that with your armadillo one liner)! I am definitely interested in learning that from you :) . I thought that the if/else statement is faster than evaluating the expression (it may occur a lot that the min or max values are present).

Reply:
Yes, you are right Ryan. I have already corrected the code on my computer a few days ago. That part looks like now this:

```
if(!is_close_to_zero((*itemMax)(i) - (*itemMin)(i)))
{
	if (val == (*itemMin)(i))
		val = scale_lower;
	else if (val == (*itemMax)(i))
		val = scale_upper;
	else
		val = scale_lower + Dx * (val - (*itemMin)(i)) / ((*itemMax)(i) - (*itemMin)(i));
}
else {
	//in case all data points are the same then set all to 1 
	val = 1.0;
}
datapoint(i) = val;
```